### DISCUSSION TITLE
Update inventory stocks

### DISCUSSION THREAD

**Discussion by ferrarinim** [abuzuhri/Amazon-SP-API-CSharp#43] [Q&A] [answered]
Ho,
First of all I love you because this library Is what I'va been searching for years! 
Now, you saved my Life.
I had managed to implement the getorders with a lot of effort and sweat.
Now, what I Need to do Is ti update stock of an item in my inventory. I understand that I have to use Feeds.
can you help me with a Little example?

Thank you very much, Happy holiday

**Comment by abuzuhri** [comment]
Hi @ferrarinim ,

Thanks :) 


Do you try this example , you can add Up to 10GB messages I thing its will be more than 200K messages

```CSharp
ConstructFeedService createDocument = new ConstructFeedService("A3J37AJU4O9RHK", "1.02");

var list = new List<PriceMessage>();
list.Add(new PriceMessage()
{
    SKU = "8201031206122...",
    StandardPrice = new StandardPrice()
    {
        currency = BaseCurrencyCode.AED.ToString(),
        Value = (201.0522M).ToString("0.00")
    }
});
createDocument.AddPriceMessage(list);

var xml = createDocument.GetXML();

var feedID = amazonConnection.Feed.SubmitFeed(xml, FeedType.POST_PRODUCT_PRICING_DATA);

Thread.Sleep(1000*30);

var feedOutput=amazonConnection.Feed.GetFeed(feedID);

var outPut=amazonConnection.Feed.GetFeedDocument(feedOutput.ResultFeedDocumentId);

var reportOutpit = outPut.Url;
```

**Comment by abuzuhri** [comment, marked as answer]
Here Update stock sample 

```CSharp
            ConstructFeedService createDocument = new ConstructFeedService("A3J37AJU4O9RHK", "1.02");
            var list = new List<InventoryMessage>();
            list.Add(new InventoryMessage()
            {
                SKU = "8201031206122...",
                Quantity = 2,
                FulfillmentLatency = "11",
            });
            createDocument.AddInventoryMessage(list);
            var xml = createDocument.GetXML();

            var feedID = amazonConnection.Feed.SubmitFeed(xml, FeedType.POST_INVENTORY_AVAILABILITY_DATA);
```

**Reply by ferrarinim** [reply]
Hi,
I tryied this code but I'm getting this error
However the qty is changed correctly.


      {
        "errors": [
          {
            "message": "Access to requested resource is denied.",
           "code": "Unauthorized",
           "details": ""
          }
        ]
      }


`          public static void UpdateStocks()
	   {

		  AmazonCredential credentials = new AmazonCredential(
			AccessKey: Creds.Default.AccessKeyId,
			SecretKey: Creds.Default.SecretKey,
			RoleArn: Creds.Default.arnRole,
			ClientId: Creds.Default.ClientId,
			ClientSecret: Creds.Default.ClientSecret,
			RefreshToken: Creds.Default.RefreshToken
		 );
		  credentials.Environment = Environments.Production;
		  credentials.MarketPlace = FikaAmazonAPI.Utils.MarketPlace.Italy;

		  AmazonConnection amazonConnection = new AmazonConnection(credentials);

		  ConstructFeedService createDocument = new ConstructFeedService(Creds.Default.VendorId, "1.02");
		  var list = new List<InventoryMessage>();
		  list.Add(new InventoryMessage()
		  {
			 SKU = "31A999100 EST-42",
			 Quantity = 2,
			 FulfillmentLatency = "1",
		  });
		  createDocument.AddInventoryMessage(list);
		  var xml = createDocument.GetXML();

		  var feedID = amazonConnection.Feed.SubmitFeed(xml, FeedType.POST_INVENTORY_AVAILABILITY_DATA);

		  Thread.Sleep(1000 * 30);

		  var feedOutput = amazonConnection.Feed.GetFeed(feedID);

		  var outPut = amazonConnection.Feed.GetFeedDocument(feedOutput.ResultFeedDocumentId);

		  var reportOutpit = outPut.Url;

	   }`

The code to get Orders works perfectly so credentials data are correct.

**Reply by abuzuhri** [reply]
> Hi, I tryied this code but I'm getting this error However the qty is changed correctly.
> 
> ```
>   {
>     "errors": [
>       {
>         "message": "Access to requested resource is denied.",
>        "code": "Unauthorized",
>        "details": ""
>       }
>     ]
>   }
> ```
> 
> ` public static void UpdateStocks() {
> 
> ```
> 	  AmazonCredential credentials = new AmazonCredential(
> 		AccessKey: Creds.Default.AccessKeyId,
> 		SecretKey: Creds.Default.SecretKey,
> 		RoleArn: Creds.Default.arnRole,
> 		ClientId: Creds.Default.ClientId,
> 		ClientSecret: Creds.Default.ClientSecret,
> 		RefreshToken: Creds.Default.RefreshToken
> 	 );
> 	  credentials.Environment = Environments.Production;
> 	  credentials.MarketPlace = FikaAmazonAPI.Utils.MarketPlace.Italy;
> 
> 	  AmazonConnection amazonConnection = new AmazonConnection(credentials);
> 
> 	  ConstructFeedService createDocument = new ConstructFeedService(Creds.Default.VendorId, "1.02");
> 	  var list = new List<InventoryMessage>();
> 	  list.Add(new InventoryMessage()
> 	  {
> 		 SKU = "31A999100 EST-42",
> 		 Quantity = 2,
> 		 FulfillmentLatency = "1",
> 	  });
> 	  createDocument.AddInventoryMessage(list);
> 	  var xml = createDocument.GetXML();
> 
> 	  var feedID = amazonConnection.Feed.SubmitFeed(xml, FeedType.POST_INVENTORY_AVAILABILITY_DATA);
> 
> 	  Thread.Sleep(1000 * 30);
> 
> 	  var feedOutput = amazonConnection.Feed.GetFeed(feedID);
> 
> 	  var outPut = amazonConnection.Feed.GetFeedDocument(feedOutput.ResultFeedDocumentId);
> 
> 	  var reportOutpit = outPut.Url;
> 
>    }`
> ```
> 
> The code to get Orders works perfectly so credentials data are correct.



is it same problem also with submit feed to change price ?

**Comment by wheelersf** [comment]
@abuzuhri  I got this work, but ended up having to use WebRequest to stream the content to a string variable.  For some reason, 'GetFeedDocumentProcessingReport' is not recognized as a member of 'amazonConnection.Feed'.  

Thanks,
Scott

               ```
                int MaxRequests = 5;
                int NumberOfRequests = 0;
                string CurrentStatus = "";
                                
                ConstructFeedService createDocument = new ConstructFeedService("SELLER_ID_HERE", "1.02");
                var list = new List<InventoryMessage>();
                list.Add(new InventoryMessage()
                {
                    SKU = "8201031206122",
                    Quantity = 2,
                    FulfillmentLatency = "11",
                });
                createDocument.AddInventoryMessage(list);
                var xml = createDocument.GetXML();

                Console.WriteLine("Submitting feed, please wait...");

                var feedID = amazonConnection.Feed.SubmitFeed(xml, FeedType.POST_INVENTORY_AVAILABILITY_DATA);

                //executes fine and status = INPROGRESS
                var feedOutput = amazonConnection.Feed.GetFeed(feedID);
                CurrentStatus = feedOutput.ProcessingStatus.ToString();

                Console.WriteLine("Initial Status = " + CurrentStatus);
 
                while (string.IsNullOrEmpty(feedOutput.ResultFeedDocumentId))
                {
                    if (NumberOfRequests == MaxRequests)
                    {
                        break;
                    }

                    NumberOfRequests += 1;

                    //Sleep 60 seconds before requesting the next status
                    Thread.Sleep(60000);

                    feedOutput = amazonConnection.Feed.GetFeed(feedID);
                                                          
                    CurrentStatus = feedOutput.ProcessingStatus.ToString();

                    Console.WriteLine("Status = " + CurrentStatus + " after check #" + NumberOfRequests.ToString());

                    if (CurrentStatus == "CANCELLED" || CurrentStatus == "FATAL")
                    {
                        break;
                    }
                }


                if (CurrentStatus == "DONE")
                {
                    var outPut = amazonConnection.Feed.GetFeedDocument(feedOutput.ResultFeedDocumentId);

                    var reportOutput = outPut.Url;
                    //Console.WriteLine(reportOutput);
                    
                    //var processingReport = amazonConnection.Feed.GetFeedDocumentProcessingReport(outPut.Url);

                    string results = "N/A";

                    try
                    {
                        HttpWebRequest req = (HttpWebRequest)WebRequest.Create(reportOutput);
                        HttpWebResponse resp = (HttpWebResponse)req.GetResponse();

                        System.IO.StreamReader sr = new System.IO.StreamReader(resp.GetResponseStream());
                        results = sr.ReadToEnd();
                        sr.Close();

                        Console.WriteLine(results);
                    }
                    catch (Exception ex)
                    {
                        results = ex.Message;
                        Console.WriteLine(results);
                    }

                }
                else
                {
                    Console.WriteLine("Wait-time for feed completion was exceeded and the final status was: " + CurrentStatus);
                }
```

**Reply by ferrarinim** [reply]
whit this code I don't have errors, thanks!

**Comment by wheelersf** [comment]
@abuzuhri While it looks like that code segment above is working, there is an issue.  The SKU that I submitted in the feed did not exist in the catalog and should have resulted in a response similar to the one below, but it didn't.  It indicated no errors at all.

Thanks,
Scott

```
<?xml version="1.0" encoding="UTF-8"?>
<AmazonEnvelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="amzn-envelope.xsd">
	<Header>
		<DocumentVersion>1.02</DocumentVersion>
		<MerchantIdentifier>SELLER_ID_HERE</MerchantIdentifier>
	</Header>
	<MessageType>ProcessingReport</MessageType>
	<Message>
		<MessageID>1</MessageID>
		<ProcessingReport>
			<DocumentTransactionID>52763017892</DocumentTransactionID>
			<StatusCode>Complete</StatusCode>
			<ProcessingSummary>
				<MessagesProcessed>1</MessagesProcessed>
				<MessagesSuccessful>0</MessagesSuccessful>
				<MessagesWithError>1</MessagesWithError>
				<MessagesWithWarning>0</MessagesWithWarning>
			</ProcessingSummary>
			<Result>
				<MessageID>1</MessageID>
				<ResultCode>Error</ResultCode>
				<ResultMessageCode>13013</ResultMessageCode>
				<ResultDescription>We&apos;re unable to complete your request because this SKU is not in the Amazon catalog. If this was a deleted SKU, wait 24 hours before resubmitting it. If you tried to add this SKU to the Amazon catalog before, check your data and correct any errors before resubmitting. </ResultDescription>
				<AdditionalInfo>
					<SKU>PURPOSELY_BAD_SKU</SKU>
				</AdditionalInfo>
			</Result>
		</ProcessingReport>
	</Message>
</AmazonEnvelope>
```

**Comment by abuzuhri** [comment]
Hi @wheelersf 

After checking this code for change price

```Csharp
ConstructFeedService createDocument = new ConstructFeedService("{SellerID}", "1.02");

            var list = new List<PriceMessage>();
            list.Add(new PriceMessage()
            {
                SKU = "820103120612.2...",
                StandardPrice = new StandardPrice()
                {
                    currency = BaseCurrencyCode.AED.ToString(),
                    Value = (201.0522M).ToString("0.00")
                }
            });
            createDocument.AddPriceMessage(list);

            var xml = createDocument.GetXML();

            var feedID = amazonConnection.Feed.SubmitFeed(xml, FeedType.POST_PRODUCT_PRICING_DATA);

            Thread.Sleep(1000 * 30);

            var feedOutput = amazonConnection.Feed.GetFeed(feedID);

            var outPut = amazonConnection.Feed.GetFeedDocument(feedOutput.ResultFeedDocumentId);
            Thread.Sleep(1000 * 30);
            var reportOutpit = outPut.Url;

            var processingReport = amazonConnection.Feed.GetFeedDocumentProcessingReport(reportOutpit);
```

I get ResultDescription

`This SKU is not in the Amazon catalog. If you are receiving this message after submitting a multi-marketplace inventory file and the designated marketplace for this error is different than the marketplace in which you submitted your file, this error is an indication that the Detail page for this item does not exist in the designated marketplace. Amazon is attempting to create the Detail Page for this item on your behalf. If successful, your listing will be created in the designated marketplace within 48 hours.`

**Reply by asifiq2000** [reply]
Hi @abuzuhri can I post the catalog feed in similar fashion as this worked for inventory and price update. 
could you help me out
one more query is if catalog cant be posted in similar way so could you help me the other way to post the 
catalog using rest api , I have already generated the catalog xml file. If you give an example or sample so it 
would help me a lot.

**Comment by wheelersf** [comment]
Hi @abuzuhri ,

That's actually a different feed than what I used.  I used the INVENTORY feed, not the PRICING feed.  The INVENTORY feed is what indicated no errors with the bad SKU.  Also, my build (1.0.29) doesn't have .GetFeedDocumentProcessingReport as a member of feed.  Was that recently added?

Anyone else experiencing what I am?

Thanks,
Scott

**Reply by abuzuhri** [reply]
after test you are true , but this problem came from amazon sp api . only update inventory give this problem .
we need to open case in amazon seller center and give them all header came from the responce when this problem happen

and for GetFeedDocumentProcessingReport  yes its new added

**Comment by astanley44** [comment]
Seems that if you call GetFeedDocumentProcessingReport (which calls GetFeedDocumentProcessingReportAsync) you get the correct result. If you call GetFeedDocumentProcessingReport directly it doesn't return the result.
Works for both POST_INVENTORY_AVAILABILITY_DATA and POST_PRODUCT_PRICING_DATA
